### PR TITLE
support query strings in checks

### LINKED ISSUE #1561: HTTP service checks should support query strings
Nomad currently takes a "path" for the URL for a consul http check.

If you attempt to include a query string, the "?" gets escaped.

### PR DESCRIPTION
**Pull request by dbresson**
resolves #1561

### CODE REVIEW

**Comment by diptanu** [comment]
@dbresson I might be wrong, but since you are using Opaque here and just setting the path, wouldn't the resultant url be something like `http://path`? 

I think we would want this -
url := url.URL{
    Scheme: check.Protocol,
    Host:   net.JoinHostPort(host, strconv.Itoa(port)),
    Path: check.Path,
    RawPath:   check.Path,
}

**Comment by dbresson** [comment]
I'm pretty sure the hostname and port are not part of the opaque section of a URI. If you're looking at the Go net/url documentation and seeing the opaque example shows the hostname as ignored, that's just because the HTTP request object has Host which overrides the host part of the URL.

**Comment by dbresson** [comment]
Ugh, I guess I'm wrong, based on this:
https://golang.org/pkg/net/url/#URL.String

**Comment by dbresson** [comment]
It looks like what we really need is URL.parse on the path, and then use URL.ResolveReference to combine the path and query with the rest of the URL.

**Comment by diptanu** [comment]
@dbresson What I suggested won't work? From the docs it feels like `RawPath` which introduced in Go 1.5 is an encoded path hint

**Comment by dbresson** [comment]
Well, it might work, but I don't think it's right. It wouldn't set the query component of the URL. It would produce an invalid URL object that just happens to produce the correct URL string.

**Comment by diptanu** [comment]
Yeah, if we use RawPath, we would also have to add the RawQuery to add the query parameters.

**Comment by dbresson** [comment]
My latest commit is probably closer to something that would work. I don't think it's quite right yet though, I suspect it will break if the path isn't set at all.

File: command/agent/consul/syncer.go [hunk 1]
```diff
@@ -703,11 +703,11 @@ func (c *Syncer) createCheckReg(check *structs.ServiceCheck, serviceReg *consul.
 		if check.Protocol == "" {
 			check.Protocol = "http"
 		}
-		url := url.URL{
+		base := url.URL{
 			Scheme: check.Protocol,
 			Host:   net.JoinHostPort(host, strconv.Itoa(port)),
-			Path:   check.Path,
 		}
+		url := base.ResolveReference(url.parse(check.Path))
```
Comment (diptanu) [inline command/agent/consul/syncer.go]: `Parse` is an exported method.

**Comment by dbresson** [comment]
Well, my test seems to pass. I also double checked and path has to have a value for http checks, so there's no problem there. I think the test failures are unrelated to my changes.

**Comment by dbresson** [comment]
@diptanu Do you have any further feedback?

**Comment by diptanu** [comment]
@dbresson Thanks! Merged.

**Pull request merged**